### PR TITLE
Show information about the journal

### LINKED ISSUE #6189: Show information about the journal
In the entry editor, there should be a button next to the journal field that displays a bunch of information about the journal in a [PopOver](https://controlsfx.bitbucket.io/org/controlsfx/control/PopOver.html). For example similar to
image
(from [eigenfactor](http://www.eigenfactor.org/projects/journalRank/rankings.php?bsearch=COMMUNICATIONS+IN+MATHEMATICAL+PHYSICS&searchby=journal&orderby=eigenfactor))

A few possible datasources are discussed here: https://academia.stackexchange.com/questions/3/where-can-i-find-the-impact-factor-for-a-given-journal (related: https://github.com/ikashnitsky/sjrdata)

### PR DESCRIPTION
**Pull request by aqurilla**
This fixes #6189 by adding a fetcher for journal information. Info buttons are added next to the `Journal` and `ISSN` fields in the entry editor, and show the information as a popover. An `EnablementStatus` enum is also added for generally maintaining state of online services in Preferences.

image

image

### Mandatory checks
- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by aqurilla** [comment]
Following the suggestion here (https://github.com/JabRef/jabref/issues/6189#issuecomment-793935314), as a first step I've written a script that combines Scimagojr data across all the available years (1999-2022). The idea is to build a consolidated json and to include it with JabRef. The data source contains info on around 38k journals.

An issue here is the size of the consolidated JSON, which is ~115MB (compressed size of 21MB), when all the fields are included for all years. These fields include - 
```
'Rank', 'Sourceid', 'Title', 'Type', 'Issn', 'SJR', 'H index', 
'Total Docs. (2020)', 'Total Docs. (3years)', 'Total Refs.', 'Total Cites (3years)', 
'Citable Docs. (3years)', 'Cites / Doc. (2years)', 'Ref. / Doc.', 'Country', 'Region', 
'Publisher', 'Coverage', 'Categories', and 'Areas'
```

Is this size acceptable or should json size be reduced further by 
a) limiting the fields included and/or
b) limiting the years included?

**Comment by Siedlerchr** [comment]
Hmm a possible solution could be to use a similar approach as the journal abbreviations, put them in a mv database.

**Comment by aqurilla** [comment]
Thanks, I'll take a look at that

**Comment by Siedlerchr** [comment]
@aqurilla some more context (was only on mobile earlier)  We are regularly updating the journal abbreviations either manually or partially automated and then merge them together to store them in a mv store database.

https://www.h2database.com/html/mvstore.html 
https://github.com/JabRef/abbrv.jabref.org 

Another option could be as a first step to download the file from a GitHub repo

**Comment by calixtus** [comment]
Is there no way to recieve the data about the journals just in time, maybe chaching them, as soon as it is questioned once? A progress indicatoor could be shown while a background process loads the journal information...

**Comment by aqurilla** [comment]
@Siedlerchr Thanks for the additional context - using an mv database would be a good option if we are going with the consolidated json approach.

@calixtus thanks for bringing this up. I tried the Elsevier API again and it looks like one of their endpoints works well for our use case (https://dev.elsevier.com/documentation/SerialTitleAPI.wadl#d1e534). It allows search by ISSN and returns the latest data for several journal metrics. We could just call the endpoint when the user clicks the journal info button. The drawback in this approach is the API quota/limits.

What would be the best approach to follow?

**Comment by calixtus** [comment]
I'm not happy with another repository distributed with every release of jabref. Adds another 30 mb to our installer. Lightweight is something different.

Maybe we can cache the journals as soon they are loaded from the API and the cache stored when closing jabref, so the API quota limit is not exceeded to soon.

**Comment by calixtus** [comment]
Also please avoid adding the future repository to a preferences object.

**Comment by tobiasdiez** [comment]
I could host the journal information on jabref online. 100mb is nothing for a real DB. Has the disadvantage that the user needs to be online.

**Comment by aqurilla** [comment]
openalex looks great w.r.t having a very high daily API quota, so there is low chance of quota exhaustion. It does seem to only have yearly counts for number of works and number of citations though

**Comment by calixtus** [comment]
## Devcall
We had a short little discussion tonight in our maintainers devcall. We really like this little enhancement, but we also see some problems that may arise. If we distribute all the journal data with the jabref package on release, it will only be update approx twice a year. Also this feature may only affect / be useful only for a very small number of users, but we would add a database of 20 mb (?) to every release. This is something we are not very happy about.
I think we should follow tobias suggestion to host the journal data on a jabref server or to fetch the journal data from another source and caching them for about a week on the user pc.
If the journal data is only updated once a year, wouln't be some valuable data be ignored? The issue #6189 suggests to show the price range and the range? This would require to download the data on the fly.
We have to keep in mind, if we make this a jit data fetcher, this has to be included in our privacy rules and made opt-in for GDPR compliance.

**Comment by calixtus** [comment]
@aqurilla thanks for your efforts already. Would be great if you could keep the concerns above in mind continuing your PR! ❤️

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
> It does seem to only have yearly counts for number of works and number of citations though

Just a quick question about this, do you have a link? I'd expect at least monthly based on 

https://docs.openalex.org/api-entities/publishers/publisher-object
- `counts_by_year`
- `cited_by_count`
- `works_count`

(perhaps you would need to combine the information to get the current years count, I haven't tried the API yet)

**Comment by aqurilla** [comment]
@calixtus thanks for sharing the devcall discussion details! Keeping all these points in mind, I think the best way would be to go ahead with a jit data fetcher using the Scopus API, and include caching on the user system. It looks like this would address all of the issues.

@k3KAW8Pnf7mkmdSMPHz27 sure! I am referring to the link that tobiasdiez shared (https://docs.openalex.org/api-entities/sources/get-a-single-source). For showing year-wise variation charts we only have the `counts_by_year.works_count` and `counts_by_year.cited_by_count` data

**Comment by aqurilla** [comment]
@tobiasdiez sounds good!

**Comment by tobiasdiez** [comment]
I now have a first draft for the API. You can try it out by issuing a POST request to https://mango-pebble-0224c3803-2067.westeurope.1.azurestaticapps.net/api with
```json
{
"query":"query GetJournalByIssn($issn: Int) {\n  journal(issn: $issn) {\n    id\n    name\n    issn\n    scimagoId\n    country\n    publisher\n    areas\n    categories\n    citationInfo {\n      year\n      docsThisYear\n      docsPrevious3Years\n      citableDocsPrevious3Years\n      citesOutgoing\n      citesOutgoingPerDoc\n      citesIncomingByRecentlyPublished\n      citesIncomingPerDocByRecentlyPublished\n      sjrIndex\n    }\n    hIndex\n  }\n}\n",
"variables":{"issn":15230864},
"operationName":"GetJournalByIssn"
}
```
This should give you a result of the type
```
{
  "data": {
    "journal": {
      "id": "ckslj3f10000f09jvc1xifgi9",
      "name": "Antioxidants & Redox Signaling",
      "issn": [
        15230864,
        15577716
      ],
      "scimagoId": 27514,
      "country": "United States",
      "publisher": "Mary Ann Liebert Inc.",
      "areas": [
        "Biochemistry, Genetics and Molecular Biology",
        "Medicine"
      ],
      "categories": [
        "Biochemistry (Q1)",
        "Cell Biology (Q1)",
        "Clinical Biochemistry (Q1)",
        "Medicine (miscellaneous) (Q1)",
        "Molecular Biology (Q1)",
        "Physiology (Q1)"
      ],
      "citationInfo": [
        {
          "year": 2022,
          "docsThisYear": 217,
          "docsPrevious3Years": 488,
          "citableDocsPrevious3Years": 487,
          "citesOutgoing": 19202,
          "citesOutgoingPerDoc": 130.63,
          "citesIncomingByRecentlyPublished": 3692,
          "citesIncomingPerDocByRecentlyPublished": 7.21,
          "sjrIndex": 1.706
        },
        {
          "year": 2021,
          "docsThisYear": 158,
          "docsPrevious3Years": 530,
          "citableDocsPrevious3Years": 530,
          "citesOutgoing": 24155,
          "citesOutgoingPerDoc": 152.88,
          "citesIncomingByRecentlyPublished": 4724,
          "citesIncomingPerDocByRecentlyPublished": 7.59,
          "sjrIndex": 1.832
        }
      ],
      "hIndex": 217
    }
  }
}
```

Currently, it only contains this particular test data. I'll later add all of the data from scimago. Let me know if you encounter any issues or questions.

**Comment by aqurilla** [comment]
@tobiasdiez thanks I will check it out. The data format looks convenient to me :+1:

**Comment by tobiasdiez** [comment]
Should be working now at https://mango-pebble-0224c3803-2067.westeurope.1.azurestaticapps.net/api.

**Comment by aqurilla** [comment]
Would it be better to use a String for the ISSN instead of an Int? because it looks like the last character can sometimes be an 'X' (https://en.wikipedia.org/wiki/ISSN#Code_format)

**Comment by tobiasdiez** [comment]
Good point, fixed!

File: src/main/java/org/jabref/gui/fieldeditors/JournalInfoOptInDialogHelper.java
Comment (koppor) [inline src/main/java/org/jabref/gui/fieldeditors/JournalInfoOptInDialogHelper.java:19]: We had a looong discussion on two settings for online services at https://github.com/koppor/jabref/issues/566.

I think, the summary is at https://github.com/JabRef/jabref/pull/9801#issuecomment-1524050565

We opted for an enum `FIRST_START, ENABLED, DISABLED`. You can introduce that generally as it will be used here and at Grobid (later by someone ^^).

File: src/main/java/org/jabref/logic/journals/JournalInformationFetcher.java
Comment (koppor) [inline src/main/java/org/jabref/logic/journals/JournalInformationFetcher.java:41]: Doesn't JabRef have an ISSN parser, similar to the DOI parser?

**Review by koppor — CHANGES_REQUESTED** [review]
Some small comments.

I tried out, but always get

image

Maybe, I should run the python file somehow?

The ISSN field should also take the (i) button. Otherwise, the UX is strang: If I push the (i) at the title, the tool says: no issn found. Then I enter an ISSN, but cannot click the button instantly.

image

**Comment by aqurilla** [comment]
Thanks for your feedback. Actually the user does not have to run the python script, since we decided to finally go with the JabRef api hosted at https://mango-pebble-0224c3803-2067.westeurope.1.azurestaticapps.net/api. I have not removed the script since it could be used for building the database behind this API.

The API was working previously with the test ISSN `15230864`, but I am also getting the following error when trying now - 
```
{
    "errors": [
        {
            "message": "\nInvalid `prisma.journal.findFirst()` invocation:\n\n\nDatabase `postgres_test_2067` does not exist on the database server at `jabrefdb.postgres.database.azure.com:5432`.",
            "locations": [
                {
                    "line": 2,
                    "column": 3
                }
            ],
            "path": [
                "journal"
            ],
            "extensions": {
                "code": "INTERNAL_SERVER_ERROR",
                "stacktrace": [
                    "Error: ",
                    "Invalid `prisma.journal.findFirst()` invocation:",
                    "",
                    "",
                    "Database `postgres_test_2067` does not exist on the database server at `jabrefdb.postgres.database.azure.com:5432`.",
                    "    at Pn.handleRequestError (C:\\home\\site\\wwwroot\\functions\\node_modules\\@prisma\\client\\runtime\\library.js:171:7205)",
                    "    at Pn.handleAndLogRequestError (C:\\home\\site\\wwwroot\\functions\\node_modules\\@prisma\\client\\runtime\\library.js:171:6358)",
                    "    at C:\\home\\site\\wwwroot\\functions\\node_modules\\@prisma\\client\\runtime\\library.js:174:2908",
                    "    at async C:\\home\\site\\wwwroot\\functions\\node_modules\\@prisma\\client\\runtime\\library.js:174:3123",
                    "    at async t._executeRequest (C:\\home\\site\\wwwroot\\functions\\node_modules\\@prisma\\client\\runtime\\library.js:174:10621)",
                    "    at async JournalService.getJournalByIssn (file:///C:/home/site/wwwroot/functions/chunks/nitro/azure.mjs:1602:18)",
                    "    at async Query.journal (file:///C:/home/site/wwwroot/functions/chunks/nitro/azure.mjs:1668:14)"
                ]
            }
        }
    ],
    "data": {
        "journal": null
    }
}
```
@tobiasdiez any idea why the test data is not accessible now?

**Comment by tobiasdiez** [comment]
Sorry, I've deleted the test-database by accident 😄 Should work again now.

File: src/main/java/org/jabref/logic/journals/JournalInformationFetcher.java
Comment (tobiasdiez) [inline src/main/java/org/jabref/logic/journals/JournalInformationFetcher.java:29]:
```suggestion
    private static final String API_URL = "https://gentle-forest-03418aa03-2067.westeurope.3.azurestaticapps.net/api";
```

**Comment by aqurilla** [comment]
No problem, thankyou! :)

File: scripts/get-journal-data.py
Comment (aqurilla) [reply scripts/get-journal-data.py:5]: yes, will do!

File: scripts/get-journal-data.py
Comment (aqurilla) [reply scripts/get-journal-data.py:6]: This should basically be the latest year of data available at https://www.scimagojr.com/journalrank.php, currently 2022. I'll add a comment in the script

File: src/main/resources/l10n/JabRef_en.properties
Comment (tobiasdiez) [reply src/main/resources/l10n/JabRef_en.properties:488]: But there are languages that translate/display the colon differently: https://en.m.wikipedia.org/wiki/Chinese_punctuation

File: src/main/resources/l10n/JabRef_en.properties
Comment (koppor) [inline src/main/resources/l10n/JabRef_en.properties:495]: Can you do title casing according to Chicago Manual of Style (or APA) here?

    Citations Outgoing per Document

File: src/main/java/org/jabref/gui/fieldeditors/JournalInfoOptInDialogHelper.java
Comment (koppor) [inline src/main/java/org/jabref/gui/fieldeditors/JournalInfoOptInDialogHelper.java:19]: I am unsure whether we should re-use existing localizations here.

```suggestion
            dialogService.notify(Localization.lang("Please enable journal information fetching in %0 > %1"), Localization.lang("Preferences"), Localization.lang("Web search"));
```

Reason: The translators change even for the same language. They might translate the strings differently (because they are "drive-by" translators; maybe students doing it as procrastination task during the time they are assigned to work on JabRef).

Comment (aqurilla) [reply src/main/java/org/jabref/gui/fieldeditors/JournalInfoOptInDialogHelper.java]: I like the term 'drive-by translators' 😄
Updated with the suggested change

File: src/main/java/org/jabref/preferences/JabRefPreferences.java
Comment (koppor) [inline src/main/java/org/jabref/preferences/JabRefPreferences.java:2941]: Ohoh, two Booleans for the same thing. -- While suggesting the tri-state, I thought, there was an easy check for key presence.

The "hacky" way would be following:

```
if (getSting(JOURNALINFO_ENABLED, "notAvailable").equals("notAvailable") {
    return EnablementStatus.FIRST_START;
} else if (getBoolean(JOURNALINFO_ENABLED)) {
  return EnablementStatus.ENABLED;
} else {
  return EnablementStatus.DISABLED;
}
```

The "clean" way could be to use the string serialization of the Enum in a String.

```` java
public enum EnablementStatus {
    FIRST_START, // The first time a user uses this service
    ENABLED,
    DISABLED;

    public static EnablementStatus fromString(String status) {
        for (EnablementStatus value : EnablementStatus.values()) {
            if (value.toString().equalsIgnoreCase(status)) {
                return value;
            }
        }
        throw new IllegalArgumentException("No enum found with value: " + status);
    }
}
```

With this way, you get rid of `JOURNALINFO_DISABLED`.

**Review by koppor — CHANGES_REQUESTED** [review]
Tried it out locally. Works!

I have an issue with the tri-state. Maybe, putting the enum as string in the preferences works better?

**Comment by tobiasdiez** [comment]
Some suggestions based on the screenshot:
- Remove the scimago id (almost nobody knows what this is and you cannot do much with it anyway)
- Move the country to the publisher column, e.g. `Mary Ann Libert (United States)`. With these changes everything should fit nicely in one row.
- Move the h-index to the first position (where currently the country is)
- Maybe use a list for the categories (each category on a new line)
- Reorder the graphs as follows:
  SJR | Cites incoming
  Cites outgoing | Cites outgoing per doc
  Docs | citable docs

File: src/main/java/org/jabref/gui/fieldeditors/PopOverUtil.java
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/fieldeditors/PopOverUtil.java:53]: The api also accepts a journal name. I think its quite common that one has the journal title but not its issn, so I would be nice to cover this situation as well.

File: src/main/java/org/jabref/preferences/JabRefPreferences.java
Comment (calixtus) [reply src/main/java/org/jabref/preferences/JabRefPreferences.java:2941]: Yes please. We also use it in other places. Go the extra mile, do it clean, do it right. :-D

File: src/main/java/org/jabref/gui/fieldeditors/PopOverUtil.java
Comment (koppor) [reply src/main/java/org/jabref/gui/fieldeditors/PopOverUtil.java:53]: Ah, nice. That means an "i" button also at the journal / journaltitle (BibLaTeX) field? This is nice!

File: src/main/java/org/jabref/preferences/JabRefPreferences.java
Comment (aqurilla) [reply src/main/java/org/jabref/preferences/JabRefPreferences.java:2941]: I was wondering about how the redundant boolean could be removed! will use this approach

File: src/main/java/org/jabref/gui/fieldeditors/PopOverUtil.java
Comment (aqurilla) [reply src/main/java/org/jabref/gui/fieldeditors/PopOverUtil.java:53]: @tobiasdiez  Thanks, will add these changes

@koppor sorry is there another journal field other than the one in 'Required fields'?

image

**Comment by calixtus** [comment]
Hi, is this ready for review?

**Comment by aqurilla** [comment]
Hi @calixtus yes this is ready, thanks for checking

**Comment by calixtus** [comment]
Sorry I wasn't able to read your PR in detail earlier, this week was unexpectedly very busy and next week isn't much better too. Please hang on...

**Comment by aqurilla** [comment]
@calixtus no worries, thanks for the update!

File: src/main/java/org/jabref/logic/journals/JournalInformationFetcher.java
Comment (koppor) [inline src/main/java/org/jabref/logic/journals/JournalInformationFetcher.java:95]: Please follow the log pattern here:

```suggestion
                    LOGGER.error("Error accessing catalog: {}", errorMessage);
```

**Review by koppor — COMMENTED** [review]
Tried it again and looked through the code. One nitpick. Nevertheless, I would vote for merging it to be able to gather more user feedback.

The only thing that worries me is that the popup shows up at another screen and is not resizable/movable here. -- I would have excpected that the popup shows up on the same screen as JabRef

grafik

**Comment by koppor** [comment]
Why is the script removed? Is it available at another repository? I would like to keep it somewhere "near" to enable switching to other endpoints.

**Comment by koppor** [comment]
Would it be ppossible to output the "INFO" only if there could not be found any information about a journal name?

```
WARN: Error while fetching journal information: ISSN and/or journal name not found in catalog: org.jabref.logic.importer.FetcherException: ISSN and/or journal name not found in catalog
        at org.jabref@100.0.0/org.jabref.logic.journals.JournalInformationFetcher.parseResponse(JournalInformationFetcher.java:129)
        at org.jabref@100.0.0/org.jabref.logic.journals.JournalInformationFetcher.getJournalInformation(JournalInformationFetcher.java:62)
        at org.jabref@100.0.0/org.jabref.gui.fieldeditors.journalinfo.JournalInfoViewModel.populateJournalInformation(JournalInfoViewModel.java:36)
        at org.jabref@100.0.0/org.jabref.gui.fieldeditors.journalinfo.JournalInfoView.populateJournalInformation(JournalInfoView.java:47)
        at org.jabref@100.0.0/org.jabref.gui.fieldeditors.PopOverUtil.lambda$showJournalInfo$0(PopOverUtil.java:40)
        at org.jabref@100.0.0/org.jabref.gui.util.BackgroundTask$1.call(BackgroundTask.java:60)
        at org.jabref@100.0.0/org.jabref.gui.util.DefaultTaskExecutor$1.call(DefaultTaskExecutor.java:161)
        at javafx.graphics@20/javafx.concurrent.Task$TaskCallable.call(Task.java:1426)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:577)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1623)
2023-07-27 00:18:22 [JavaFX Application Thread] org.jabref.gui.JabRefDialogService.notify()
INFO: Error while fetching journal information: ISSN and/or journal name not found in catalog
```

**Comment by tobiasdiez** [comment]
I've removed the python script here and will add it to https://github.com/JabRef/JabRefOnline/pull/2067. Also updated the api url to the production server. This needs https://github.com/JabRef/JabRefOnline/pull/2067 to be merged, which I will hopefully do tomorrow or latest at the weekend.

File: src/main/java/org/jabref/logic/journals/JournalInformationFetcher.java
Comment (aqurilla) [reply src/main/java/org/jabref/logic/journals/JournalInformationFetcher.java:95]: Hmm thats strange, the popup is coded to open at the point of the button. I'll check this once the api PR is merged

**Comment by tobiasdiez** [comment]
I've now merged the PR in the jabref online repo, but the data ingestion is still ongoing (might take a few hours). You can already test it with issn 15454509.

File: src/main/java/org/jabref/logic/journals/JournalInformationFetcher.java
Comment (koppor) [inline src/main/java/org/jabref/logic/journals/JournalInformationFetcher.java:33]: @tobiasdiez We would like to separate the http representation from the api functionality. Can that also be hosted at `api.jabref.org`?

File: src/main/java/org/jabref/logic/journals/JournalInformationFetcher.java
Comment (koppor) [inline src/main/java/org/jabref/logic/journals/JournalInformationFetcher.java:53]: @tobiasdiez I did not see that during the first reviews. This is Level 0 in the [Richardson Maturity Model](https://martinfowler.com/articles/richardsonMaturityModel.html). Maybe for now, this is OK. Regarding the aim of education of properly designed APIs, we might consider an udpate on the http resource design. -- I think, this is a longer topic and does not need to be solved now. It is more an input for https://github.com/JabRef/jabref/pull/9748.

Comment (tobiasdiez) [reply src/main/java/org/jabref/logic/journals/JournalInformationFetcher.java:53]: Because all the api endpoints are graphql endpoints and there is no point in making an exception for the journal infos. Now that jabref no longer has the library limits, we can work on using a proper java graphql client but for this PR I think its good enough to use http + manual parsing (which is actually not that different than using a rest call)

File: src/main/java/org/jabref/logic/journals/JournalInformationFetcher.java
Comment (tobiasdiez) [reply src/main/java/org/jabref/logic/journals/JournalInformationFetcher.java:33]: Not really, both the website and in the future the office addin needs the `/api` endpoint, see https://github.com/JabRef/JabRefOnline/issues/576. I was reserving `api.jabref.org` for a potential endpoint for external services (e.g. if other people/apps want to call our api) but that requires more thought and work.

**Review by tobiasdiez — APPROVED** [review]
Can we get this in as soon as possible? It's a very nice new feature and it would be nice to get some feedback from dev build users before the next release (mainly because after the release I need to be careful with api changes to not break backwards compatibility).

**Comment by aqurilla** [comment]
Hi, is there any further work required for this PR? It looks like the failing test is unrelated

**Comment by tobiasdiez** [comment]
I've merged this now as there were no further requests for changes. Thanks a lot @aqurilla for your nice work on this, and your patience!

**Comment by aqurilla** [comment]
Thankyou! appreciate the extensive discussions

**Pull request merged**
